LibSQL: Data Storage

JanDeVisser opened the pull request:
This is the beginning of the LibSQL data storage functionality. Data is stored in a Heap object, which is a block based file.
BTrees (and their constituent nodes and keys) and Tuples can be stored in a Heap. A very simple Database object manages
rudimentary SQL tables in a Heap.

With apologies for the monster patch. I'm new to this github stuff and the way this project works, but I promise I will learn...

Hendiadyoin1 commented on Userland/Libraries/LibSQL/TreeNode.cpp:
```diff
@@ -0,0 +1,284 @@
+Optional<u32> TreeNode::get(Key& key)
+{
+    dbgln("[{}] find({})",
+        m_pointer, (String) key);
```
is there not a better way than to c-cast it?

JanDeVisser replied:
Sure. I will create a to_string() method and have the (String) operator use that.

Hendiadyoin1 replied:
You don't need to override the operator nor use a c-cast here if you make a to_string() function, just call it directly

Hendiadyoin1 commented on Userland/Libraries/LibSQL/TreeNode.cpp:
```diff
@@ -0,0 +1,284 @@
+    return down_node(size())->node_for(key);
+}
+
+Optional<u32> TreeNode::get(Key& key)
```
is key supposed to cahnge in this function?

if yes:
dont use a reference
if no: 
mark it as constant

I dont think a getter should change the value of the provided index/key

JanDeVisser replied:
Hrm. It doesn't change the value per se, it sets the pointer it finds for that key in the key. So if the key is present in the node, the return value is set in the key as well.

More that willing to change name or semantics, just not sure where to go.

Hendiadyoin1 replied:
again seems weir for a `get(...)`, but OK

Hendiadyoin1 commented on Userland/Libraries/LibSQL/TreeNode.cpp:
```diff
@@ -0,0 +1,284 @@
+        serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[ix]);
+        entry.serialize(buffer);
+    }
+    serialize_to(buffer, is_leaf() ? (u32) 0 : m_down_pointers[size()]);
```
you can just append a `u` to a number to make it unsigned, like `0u`

JanDeVisser replied:
Done

Hendiadyoin1 commented on Userland/Libraries/LibSQL/TreeNode.cpp:
```diff
@@ -0,0 +1,284 @@
+        builder.append(" is leaf");
+    } else {
+        builder.appendff(" down nodes {}", m_down_nodes.size());
+    }
```
when all paths of the comparison are just one line, we usually omit the curly braces

you might have to fix it on earlyer lines too, this is the first time that I conciously see it

JanDeVisser replied:
30 years of habit don't go away that easily, young padawan :-)

I know there's discussions bordering on holy wars on what's best here; I've always used braces but it's not worth the aggravation. I'll remove them.

Hendiadyoin1 commented on Userland/Libraries/LibSQL/Tuple.cpp:
```diff
@@ -0,0 +1,13 @@
+#include <LibSQL/Tuple.h>
+
+namespace SQL {
+
```
empty?

JanDeVisser replied:
Not anymore.

Hendiadyoin1 commented on Userland/Libraries/LibSQL/Type.h:
```diff
@@ -0,0 +1,25 @@
+#define __ENUMERATE_SQL_TYPE(name, type, impl) type,
+    ENUMERATE_SQL_TYPES
+#undef __ENUMERATE_SQL_TYPE
+};
```
this looks weird

JanDeVisser replied:
Isn't this the pattern Andreas uses?

linusg replied:
yes, it is. not necessarily weird, but haven't looked at it in detail yet.

Hendiadyoin1 replied:
similar, but not quite
see 
https://github.com/SerenityOS/serenity/blob/master/Kernel/API/Syscall.h#L31

so it would be
```cpp
#define ENUMERATE_SQL_TYPES(S)   \
     S("text", Text, String)                     \
     S("int", Integer, int)                        \
     S("float", Float, double)

enum SQLType {
#define __ENUMERATE_SQL_TYPE(name, type, impl) type,
    ENUMERATE_SQL_TYPES(__ENUMERATE_SQL_TYPE)
#undef __ENUMERATE_SQL_TYPE
};

```

JanDeVisser replied:
Ah OK will fix

Hendiadyoin1 commented on Userland/Libraries/LibSQL/Meta.cpp:
```diff
@@ -10,78 +10,112 @@
 
-ColumnDef::ColumnDef(String const& name, SQLType sql_type)
-    : Relation(name)
+/* ----------------------------------------------------------------------- */
```
why do you use those lines?

JanDeVisser replied:
I use them to find my way around my source. I will delete them

Hendiadyoin1 commented on Userland/Libraries/LibSQL/TreeNode.cpp:
```diff
@@ -94,15 +97,21 @@ bool TreeNode::insert(Key const& key, TreeNode *right)
 
+#ifdef BTREE_DEBUG
     dbgln("Inserting in this node");
+#endif
```
there is something called `dbgln_if`
it essentially does the same, but looks neater and is the preferred way of doing this

Hendiadyoin1 commented on Tests/LibSQL/TestSQLStorage.cpp:
```diff
@@ -100,6 +101,194 @@ TEST_CASE(create_heap) {
+    }
+}
+
+TEST_CASE(btree_one_key)
```
why do you make 10 tests doing essentially the same?

JanDeVisser replied:
Because it is my experience that with datastructures like this it's impossible to design your test so you find all corner cases. Doing it this way is a bit of a blunt force instrument but I've had good experiences with it, including in this code.

Hendiadyoin1 commented on Tests/LibSQL/TestSQLStorage.cpp:
```diff
@@ -100,6 +101,194 @@ TEST_CASE(create_heap) {
+    insert_into_and_scan_btree(50);
+}
+
+#if 0
```
whats with this unused code

Hendiadyoin1 reviewed (commented):
This is a very shallow review for now, and I stoped like half way through, but there are some things that should definetly be changed

please cleanup your git history:
there are a lot of commits that essentially say "Test and fix" and you create files, just to delete them 2 commits later
in addition to that the commits arent very atomic, you *test and fix* instead of `fix` then `test to prevent regression` like it is usually done
To the point I have read your history should look something like:
1. LibSQL: Add BTree
2. LibSQL: Add Storage functionality
3. Tests+LibSQL:  Add Tests

lastly please add newlines at the end of every file

alimpfard reviewed (changes requested):
This seems like a nice addition, but please observe the contribution guidelines and code style requirements, and clean up the commit history:
https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md

JanDeVisser commented on Userland/Libraries/LibSQL/BTree.h:
```diff
@@ -0,0 +1,293 @@
+}
+
+template <typename KeyType>
+int deserialize(TreeNode<KeyType> &leaf, ByteBuffer const &buffer, int offset)
```
Thanks- will look into.

JanDeVisser replied:
reading again I now see you probably meant 'tree' instead of 'free', and in that case let me explain. The tree lives in persistent storage. As you traverse it to find keys or iterate over all keys, you need to load nodes from storage, and if you insert keys, update what they point to, or delete keys, you need flush the new tree structure to storage.

This is done using the serialization and deserialization methods.

alimpfard replied:
I actually meant "free" - as in not part of a specific class.

JanDeVisser replied:
Ah OK. They're mostly gone; I only have template overrides for basic types now.

JanDeVisser commented on Userland/Libraries/LibSQL/Value.h:
```diff
@@ -71,6 +72,7 @@ class Value {
     Function<void(ByteBuffer&, size_t &offset)> m_deserialize;
     Function<size_t()> m_size;
     Function<const char *()> m_type_name;
+    Function<bool(Value const&)> m_can_cast;
```
Yes. I really wanted to be able to pass Values by reference as stack objects and be able to store them in containers as objects. If I would have used virtual methods this would have been impossible.

JanDeVisser commented on Userland/Libraries/LibSQL/ReadMe.md:
```diff
@@ -0,0 +1,70 @@
+### SQL Storage Overview
+
```
OK

JanDeVisser commented on Userland/Libraries/LibSQL/Value.cpp:
```diff
@@ -313,6 +313,7 @@ void Value::setup_text()
     };
 
     m_can_cast = [](Value const& other) -> bool {
+        (void) other;
```
OK

JanDeVisser commented on Userland/Libraries/LibSQL/BTree.cpp:
```diff
@@ -0,0 +1,7 @@
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include "BTree.h"
```
Thanks- as I mention below my git/github skills leave to be desired, I'll attempt to clean stuff up so you guys don't have to look at things that have been gone for weeks

JanDeVisser commented on Userland/Libraries/LibSQL/BTree.h:
```diff
@@ -36,7 +37,7 @@ class BlockFile {
         if (!(*m_file).seek(block * BLOCKSIZE))
         {
-            warnln("Could not seek block %d from file %s", block, m_name);
+//            warnln("Could not seek block %d from file %s", block, m_name);
```
I was unaware of this. Will investigate but I don't doubt I can use it.

JanDeVisser commented on Userland/Libraries/LibSQL/BTree.h:
```diff
@@ -131,162 +148,336 @@ class TreeNodeEntry {
     u32 m_pointer;
     Vector<TreeNodeEntry<KeyType>> m_entries;
     Vector<u32> m_down_pointers;
     Vector<TreeNode<KeyType> *> m_down_nodes;
```
Later revisions use smart pointers.

JanDeVisser commented on Userland/Libraries/LibSQL/BTree.h:
```diff
@@ -6,479 +6,124 @@
 };
 
-}
+} // namespace SQL
```
Will remove. I use it so I don't get confused when I'm missing closing braces.

JanDeVisser commented on Userland/Libraries/LibSQL/CMakeLists.txt:
```diff
@@ -2,7 +2,16 @@ set(SOURCES
+    Meta.cpp
+    TreeNode.cpp
+    Tuple.cpp
+    Value.cpp
```
Will fix

JanDeVisser commented on Userland/Libraries/LibSQL/Key.cpp:
```diff
@@ -0,0 +1,172 @@
+
+namespace SQL {
+
+Key::Key()
```
I'm not sure what the guideline is as to what can go in the header and what not. I basically moved everything out-of-line, but if there's stuff that should go inline I'm more than willing to move.

Hendiadyoin1 replied:
as long as it isn't dependent on functions after it, it usually is inline

when the functions get to long they are moved out of line too

Hendiadyoin1 commented on Userland/Libraries/LibSQL/TreeNode.cpp:
```diff
@@ -0,0 +1,284 @@
+    if (!m_tree->duplicates_allowed()) {
+        for (auto& entry : m_entries) {
+            if (key == entry) {
+                dbgln("BTree does not allow duplicates and key {} already exists");
```
how did this compile, its missing an argument to fill in the fromat

Hendiadyoin1 replied:
Not as far as I know, would need to ask on discord myself...

alimpfard replied:
Is your host compiler clang?
The format checks aren't enabled on clang

JanDeVisser replied:
gcc 10.2

JanDeVisser commented on Userland/Libraries/LibSQL/BTree.h:
```diff
@@ -0,0 +1,293 @@
+
+template <typename KeyType>
+class BTree {
+public:
```
Maybe, but in this particular case integration in the actual storage is important. What I mean by that is that the tree lives in persistent storage and only the parts needed for the job at hand should be loaded into memory (it doesn't do a very good jib at that yet but that's beside the point). Also, the keys don't refer to a living, breathing C++ object, but to another piece of data of arbitrary type also living in persistent storage.

So while it may be possible to come up with something that does double duty satisfying these requirements and being a general-purpose in-memory data structure, that's not the problem I'm trying to solve here.

alimpfard replied:
Sure.

JanDeVisser commented on Userland/Libraries/LibSQL/BTree.cpp:
```diff
@@ -40,37 +45,56 @@ BTreeIterator BTree::end()
+        auto buffer = read_block(pointer);
+        size_t offset = 0;
+        m_root = make<TreeNode>(*this, nullptr, pointer, buffer, offset);
+        m_pointer = pointer;
```
Yeah, I pondered what to call it, and in the end I settled on 'pointer' even though it's a horribly overloaded term. What it is is an address in persistent storage the key 'points' to. Other possible names would be things like 'record number', 'block number', etc. None of the alternatives seemed general enough for what I wanted to express so I settled on 'pointer',

JanDeVisser commented on Userland/Libraries/LibSQL/Key.cpp:
```diff
@@ -40,24 +40,33 @@ Key::Key(IndexDef const& index, ByteBuffer& buffer, size_t& offset)
-//        auto &key_part_definition = m_part_definitions[ix];
-//        dbgln("Serializing key[{}] = {}", key_part_definition.name(), (str_opt.has_value()) ? str_opt.value() : "(null)");
+    for (auto &key_part : m_key_data) {
+#ifdef SERIALIZE_DEBUG
```
I was pointed towards the dbgln_if functionality,

JanDeVisser commented:
I think I messed it up even more now. I rebased from the last commit before my first commit, and squashed everything into one commit. But it dragged in all other commits because I synced my tree every now and then.

I'm going to close this and start over.

Big learning experience here. Apologies you have to be part of it.
